### ISSUE TITLE
ProtocolClient Interface does not allow to specify the propertyName, actionName or eventName, if the information is not contained in the Form

### ISSUE THREAD

**Issue by RobWin** [eclipse-thingweb/node-wot#1342]
The implementation of the ProtocolClient methods in node-wot assume that all the information needed to read a property, invoke an action or subscribe to an event is contained in a Form. See https://github.com/eclipse-thingweb/node-wot/blob/master/packages/core/src/protocol-interfaces.ts#L48
When implementing a websocket binding based on the Webthingprotocol, we saw that this is not the case for all protocols. 
We had a long discussion on Discord about it. And in the Webthingprotocol spec the form href will not contain a property name or action name, so we would need to overload ProtocolClient methods with additional versions which also take the names as an input parameter.

**Comment by danielpeintner** [comment]
It seems that the `readResource()` / `writeResource()` / `invokeResource()` / `invokeResource()` / `unlinkResource()` method calls could be extended to hand in the interaction name.

Having said that, I wonder what is the best way forward with the following aspects in mind:
* most bindings don't need it..
* failure cases: property names might be different.. What should happen?
* backward compatibility  (adding and optional argument or require this new argument)
  - if optional, failing if needed?
  - more aspects? 
* is the name of an interaction enough, or might we need to add more (maybe in the future?) --> adding an (optional) object instead?


Any opinions?

**Comment by RobWin** [comment]
In [kotlin-wot](https://github.com/RobWin/kotlin-wot) I modified `ConsumedThing.readProperty` to call `client.readResource(affordanceName, form)`. 
In the ProtocolClient I kept added a new overloaded method with a default implementation. Thus I didn't have to modify existing bindings, but can override the new method in the WebsocketProtocolClient. 

```
  suspend fun readResource(resourceName: String, form: Form): Content {
      return readResource(form)
  }
  ```
  
 But  Typescript cannot have default implementations in interfaces or?

**Comment by JKRhb** [comment]
In dart_wot, I actually had a similar problem and solved it (for now) using an additional [`AugmentedForm`](https://github.com/eclipse-thingweb/dart_wot/blob/b656efea4da51a936dbcb42feb9eed89363824d0/lib/src/core/implementation/augmented_form.dart) class, that contains references to both the related interaction affordance and Thing Description. Having a best practice here across implementations would definitely be nice.

**Comment by RobWin** [comment]
Indeed, we need an object, because the `thingId` is also required in the WebthingProtocol websocket messages.

**Comment by danielpeintner** [comment]
@relu91 do you have additional opinions?

I kind of like the proposal from @JKRhb to hand over an augmented form.
I do see the rationale for the "interaction affordance" part. I wonder why you needed the full TD? @JKRhb do you need parts outside the interaction? In node-wot it shouldn't be needed (I think) since the `href` etc. is converted to absolute hrefs... or do I miss anything?

**Comment by RobWin** [comment]
I'm not sure I understand you correctly. There are protocols where the current form schema and href might not be sufficient to describe everything needed for communication.

Do you mean that a protocol binding can extend the form by adding protocol-specific extensions?

**Comment by relu91** [comment]
Let's try with a possible example. In theory, we should have a protocol binding document for Webthing protocol. Let's assume that in there they define that to describe the configuration in forms you **must** use `wtv:affordanceName`. Therefore, all valid forms that uses the webthing protocol binding should look like:
```js
{
  "href": "https://websocket.endpoint"
  "wtv:affordanceName": "propertyA"
}
```

If people are ok with this we don't have to do anything as the Form instance contains all the information needed. However, one of the most common complains for ThingDescriptions is that it is too verbose. Therefore, the protocol binding document could say that `wtv:affordanceName` is optional and its default is equal to the name of the affordance where the form is defined. 

To support this default filling we can preprocess the TD and ask the protocol binding implementation (which knows the defaults rules) to fill the missing fields. I know that in the end the result is that the protocol binding would need to access application level information, but at least we separate the concerns and make clear that the form is the only source when processing a WoT operation request.

**Comment by danielpeintner** [comment]
To make it a bit more practical ... which might help to understand what we propose to do for node-wot (and maybe other implementation can do the same?).

Our assumption is that handing in a form for a read request should be sufficient (just like we do today in https://github.com/eclipse-thingweb/node-wot/blob/d534535d68d2583a04e5d930a9aeba28e94cef14/packages/core/src/protocol-interfaces.ts#L50C18-L50C28).

If a given protocol needs additional information (such as the name or any other information) it can be added in a protocol-specific pre-processing step. This step is protocol dependent and might even be a void operation in most of the cases.

@RobWin let's get back to the example you raised about the need to know the interaction name.

We plan to extend `ProtocolClient` with something like the following

```
export interface ProtocolClient {
    /** update the form with additional information (if needed) */
    enhanceForm(form: Form, ...): Promise<void>;
}
```

node-wot will go over the forms and pass it to the according protocol implementation and ask each binding to add additional information (if needed). As said, in most of the cases it will be a no-op. In the case where the name is needed, you can add this information to the form.

A subsequent `readResource(form: Form)` will hand in the form you enhanced before. So it is up to each binding to do a proper job.

I hope this explains a bit our current thinking.
Note: at the moment it is just an idea, and we don't know yet how the *enhancement* process might look like exactly.

Does this make sense?

**Comment by JKRhb** [comment]
> We’re still interested in hearing from @JKRhb about the specific need for passing additional context information to the protocol interface.

I now finally found the time to revisit my implementation and the reason I added the `AugmentedForm` class and I think the main context information that motivated was actually the TD's ID which I intended to use in callbacks for passing in credentials to the binding implementation. However, I noticed that this information is actually not really necessary or that it is more useful to map the credentials to host names/base URLs.

A more important aspect in practice, though, is the exposure of `securityDefinitions` to the binding implementations, which makes it unnecessary to have a "setSecurity" method on the protocol interfaces, making the interaction methods act solely on the fields of the form that has been passed. Other aspects I resolve are the `href`, `security`, and `uriVariables`, but this could also be done via a constructor that accepts a `Form` and `ThingDescription`. In the case of the security definitions, however, an extension of the original form class is currently needed.
